packages/sdk/src/types/json.ts

```diff
@@ -0,0 +1,329 @@
+            );
+            break;
+    }
+
+    if (newValue !== undefined) {
+        return { transformed: true, value: newValue };
+    }
+
+    return { transformed: false, value };
+}
+
+type ReplacerFun = (this: any, key: string, value: any) => any;
+
+/**
+ * Thrown if a circular reference is found while trying to stringify object.
+ */
+export class JsonCircularReferenceError extends Error {
+    public override readonly name = 'CircularReferenceError';
+    /**
+     * @param {string} key - The key the circular reference was found at.
+     */
+    constructor(public readonly key: string) {
+        super(`Circular reference found in object at path ${key}`);
+    }
+}
+
+/**
+ * Creates a replacer function which is a no-op, but throws {@link JsonCircularReferenceError} when finding circular references.
+ * Modified from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value.
+ *
+ * @throws {JsonCircularReferenceError} If a cyclic reference is found.
+ * @returns {ReplacerFun} replacer function, which throws when finding circular references.
+ * The function returned expects parent object to be accessible on `this`
+ *
+ * @example
+ * const check = getCheckCircular();
+ *
+ * const circularReference = { otherData: 123 };
+ * circularReference.myself = circularReference;
+ *
+ * for (cosnt key in circularReference) {
+ *   check.call(circularReference, key, circularReference[key]); // throws `JsonCircularReferenceError`
+ * }
+ */
+function getCheckCircular(): ReplacerFun {
+    const ancestors: any[] = [];
+    return function (this: any, key: string, value: any) {
+        if (typeof value !== 'object' || value === null) {
+            return value;
+        }
+        // `this` is the object that value is contained in,
+        // i.e., its direct parent.
+        while (ancestors.length > 0 && ancestors.at(-1) !== this) {
+            ancestors.pop();
+        }
+        if (ancestors.includes(value)) {
+            throw new JsonCircularReferenceError(key);
+        }
+        ancestors.push(value);
+        return value;
+    };
+}
+
+/**
+ * Transforms concordium domain types in an object of arbitrary depth in a non-recursive manner.
+ *
+ * @param {unknown} obj - The object to transform
+ * @param {ReplacerFun} [replacer] - An optional replacer function to run in addition to transforming concordium domain types.
+ *
+ * @throws {JsonCircularReferenceError} If a circular reference is found.
+ * @returns {any} The transformed object.
+ */
+function transformConcordiumTypes(obj: unknown, replacer?: ReplacerFun): any {
+    if (typeof obj !== 'object' || obj === null) {
+        return obj;
+    }
+
+    type StackItem = {
+        obj: Record<string, any>;
+        path: string[];
+    };
+
+    const checkCircular = getCheckCircular();
+
+    const stack: StackItem[] = [{ obj, path: [] }];
+    const result: Record<string, any> = { ...obj }; // Shallow clone to aviod modifying original.
```

Comment:
You never read from this before assigning the new values, so no need to shallow clone
```suggestion
    const result: Record<string, any> = {};
```